app/models/runs_user_reducers.rb

```diff
@@ -0,0 +1,82 @@
+class RunsUserReducers
+  class ReductionConflict < StandardError; end
+
+  attr_reader :reducible, :reducers
+
+  def initialize(reducible, reducers)
+    @reducible = reducible
+    @reducers = reducers
+  end
+
+  def reduce(user_id, extract_ids=[], and_check_rules: false)
+    return [] unless reducers&.present?
+    return [] unless user_id
+    retries ||= 2
+
+    extract_query = prepare_extract_query(user_id)
+    extracts = FetchExtractsByUser.for(reducers).extracts(extract_query, extract_ids)
+
+    reduction_filter = { reducible_id: reducible.id, reducible_type: reducible.class.to_s, user_id: user_id }
+    reduction_fetcher = ReductionFetcher.new(reduction_filter)
+
+    # prefetch all reductions to avoid race conditions with optimistic locking
+    if reducers.any?{ |reducer| reducer.running_reduction? }
+      reduction_fetcher.load!
+    end
+
+    new_reductions = reducers.map do |reducer|
+      reducer.process(
+        extracts, 
+        reduction_fetcher.for!(reducer.topic), 
```

Comment:
Minor: I still find this confusing, it could be the fetcher is updated with the topic / strategy before we use it here. 
```
  reduction_fetcher.set_topic(reducer.topic)
...
  reducer.process(
    extracts,
    reduction_fetcher,
    relevant_reductions(extracts, reducer)
  )
```

Reply:
Totally agree, I'd just missed this. Will make this go away.

Reply:
And there's no point for it anyway, since we already know that over here we've only got reducers for Users, and same for the RunsSubjectReducers